src/widget_chooser.c

```diff
@@ -0,0 +1,502 @@
+					gtk_file_filter_set_name(filter, mime_types->line[count]);
+					gtk_file_filter_add_mime_type(filter, mime_types->line[count]);
+					gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(var->Widget), filter);
+				}
+				if (mime_types) list_t_free(mime_types);
+			}
+
+			/* File filters of type pattern are listed after mime-type filters */
+			if ((tagattr_value = get_tag_attribute(var->widget_tag_attr, "fs-filters"))) {
+				patterns = linecutter(g_strdup(tagattr_value), '|');
+				for (count = 0; count < patterns->n_lines; count++) {
+					filter = gtk_file_filter_new();
+					gtk_file_filter_set_name(filter, patterns->line[count]);
+					gtk_file_filter_add_pattern(filter, patterns->line[count]);
+					gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(var->Widget), filter);
+				}
+				if (patterns) list_t_free(patterns);
+			}
+		}
+
+		/* Connect signals */
+
+		g_signal_connect(G_OBJECT(var->Widget), "file-activated",
+			G_CALLBACK(on_any_widget_file_activated_event), (gpointer)var->Attributes);
+		g_signal_connect(G_OBJECT(var->Widget), "selection-changed",
+			G_CALLBACK(on_chooser_widget_selection_changed_event), (gpointer)var->Attributes);
+		g_signal_connect(G_OBJECT(var->Widget), "update-preview",
+			G_CALLBACK(on_chooser_widget_update_preview_event), (gpointer)var->Attributes);
+		g_signal_connect(G_OBJECT(var->Widget), "current-folder-changed",
+			G_CALLBACK(on_any_widget_current_folder_changed_event), (gpointer)var->Attributes);
+		/* step: confirm-overwrite not implemented */
+	}
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Removeselected                                                      *
+ ***********************************************************************/
+
+void widget_chooser_removeselected(variable *var)
+{
+	gchar            *var1;
+	gint              var2;
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	fprintf(stderr, "%s(): Removeselected not implemented for this widget.\n",
+		__func__);
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Save                                                                *
+ ***********************************************************************/
+
+void widget_chooser_save(variable *var)
+{
+	FILE             *outfile;
+	GList            *element;
+	GtkTextBuffer    *buffer;
+	GtkTextIter       start, end;
+	gchar            *act;
+	gchar            *filename = NULL;
+	gchar            *text;
+	gchar            *string;
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	/* We'll use the output file filename if available */
+	act = attributeset_get_first(&element, var->Attributes, ATTR_OUTPUT);
+	while (act) {
+		if (strncasecmp(act, "file:", 5) == 0 && strlen(act) > 5) {
```

Comment:
It's more efficient to check the string length first. strlen() scans the string at O(n), but checks only one condition for each byte. strncasecmp() is also O(n), but has more than one `if` per byte. And, the second condition sounds more likely to be false.

Reply:
You're correct. I just reused the same code lines that gtkdialog uses for every other widget.  Maybe in the future we should improve all widget and refactor code: there's a lot of repetition, probably due to the venerable age of gtkdialog and the very long maintenance hiatus.

Reply:
Should I revert this? I think not yet, as the code does need updating - a big job. 

Perhaps a separate repo needs to be created, away from the 'puppy' brand, even away from github. Maybe then it *could* attract more interest.

Thoughts?

Reply:
> Should I revert this

Nope, everything can be fixed later :)

Reply:
Thank you @dimkr and @01micko for merging this PR!

Reply:
> Perhaps a separate repo needs to be created, away from the 'puppy' brand, even away from github. Maybe then it _could_ attract more interest.
> 
> Thoughts?

My thoughts are that a separate repo would further dilute interest in gtkdialog development because it would look like yet another underpowered fork.  Besides, if gtkdialog is part of woof-CE any new development will get a chance at being field-tested in the next new puplet. A separate repo wouldn't get such a chance.

Reply:
I think I tend to agree to some extent. I don't think the repo's location or ownership matter much, as long as the work gets done.

gtkdialog is crucial to Puppy's survival, and the most important things are 1) people who aren't afraid to improve it and fix issues and 2) real-world testing of gtkdialog, which can happen only if woof-CE builds the latest gtkdialog and people use the built Puppy.

(And I think step- earned a spot in the puppylinux-woof-CE organization.)

Reply:
> (And I think step- earned a spot in the puppylinux-woof-CE organization.)

I can arrange that if @step- so desires.

Reply:
@01micko @dimkr  yes. To set the right expectation, daily I use Fatdog therefore I think I can contribute to woof-CE where Puppy and Fatdog overlap so I can reuse my knowledge.